bytecomp/matching.ml

```diff
@@ -937,91 +1078,40 @@ let pat_as_constr = function
+and group_var p =
+  match Pattern_head.desc (Simple.head p) with
+  | Any -> true
   | _ -> false
 
-and group_constructor = function
-  | { pat_desc = Tpat_construct (_, _, _) } -> true
-  | _ -> false
-
-and group_same_constructor tag = function
-  | { pat_desc = Tpat_construct (_, cstr, _) } ->
-      Types.equal_tag tag cstr.cstr_tag
-  | _ -> false
-
-and group_variant = function
-  | { pat_desc = Tpat_variant (_, _, _) } -> true
-  | _ -> false
-
-and group_var = function
-  | { pat_desc = Tpat_any } -> true
-  | _ -> false
-
-and group_tuple = function
-  | { pat_desc = Tpat_tuple _ | Tpat_any } -> true
-  | _ -> false
-
-and group_record = function
-  | { pat_desc = Tpat_record _ | Tpat_any } -> true
-  | _ -> false
-
-and group_array = function
-  | { pat_desc = Tpat_array _ } -> true
-  | _ -> false
-
-and group_lazy = function
-  | { pat_desc = Tpat_lazy _ } -> true
-  | _ -> false
-
-let can_group p =
-  match p.pat_desc with
-  | Tpat_any -> group_var
-  | Tpat_constant (Const_int _) -> group_const_int
-  | Tpat_constant (Const_char _) -> group_const_char
-  | Tpat_constant (Const_string _) -> group_const_string
-  | Tpat_constant (Const_float _) -> group_const_float
-  | Tpat_constant (Const_int32 _) -> group_const_int32
-  | Tpat_constant (Const_int64 _) -> group_const_int64
-  | Tpat_constant (Const_nativeint _) -> group_const_nativeint
-  | Tpat_construct (_, { cstr_tag = Cstr_extension _ as t }, _) ->
+let can_group group_discr row_pat =
+  let discr_head = Simple.head group_discr in
+  let row_head = Simple.head row_pat in
+  match Pattern_head.desc discr_head, Pattern_head.desc row_head with
+  | Any, Any
+  | Constant (Const_int _), Constant (Const_int _)
+  | Constant (Const_char _), Constant (Const_char _)
+  | Constant (Const_string _), Constant (Const_string _)
+  | Constant (Const_float _), Constant (Const_float _)
+  | Constant (Const_int32 _), Constant (Const_int32 _)
+  | Constant (Const_int64 _), Constant (Const_int64 _)
+  | Constant (Const_nativeint _), Constant (Const_nativeint _) ->
+      true
+  | Construct { cstr_tag = Cstr_extension _ as discr_tag },
+    Construct row_cstr ->
       (* Extension constructors with distinct names may be equal thanks to
          constructor rebinding. So we need to produce a specialized
          submatrix for each syntactically-distinct constructor (with a threading
          of exits such that each submatrix falls back to the
          potentially-compatible submatrices below it).  *)
-      group_same_constructor t
-  | Tpat_construct _ -> group_constructor
-  | Tpat_tuple _ -> group_tuple
-  | Tpat_record _ -> group_record
-  | Tpat_array _ -> group_array
-  | Tpat_variant (_, _, _) -> group_variant
-  | Tpat_lazy _ -> group_lazy
-  | _ -> fatal_error "Matching.can_group"
+      Types.equal_tag discr_tag row_cstr.cstr_tag
+  | Construct _, Construct _
+  | Tuple _, (Tuple _ | Any)
+  | Record _, (Record _ | Any)
```

Comment:
If like me you wonder why these allow any, my rational was "well, we're going to call `divide_line` on these groups, as we do for variables, so ...".
But:
- we're also calling `divide_line` on lazy groups, and they don't benefit from that treatment
- it's not symmetric: we could imagine `| Any, (Any | Tuple _ | Record _)`... but no. If both you and Luc happen to be at gallium today or tomorrow, perhaps you could ask him.

Reply:
After much thinking I convinced myself that it was *more correct* to not allow `Any` here, and that it may even be wrong to allow `Any`. Shortly (see a full comment for the full review): the reasoning to add Any to the other groups is that, at the types involved, Any is exactly equivalent to a more refined head that falls within the group. This is not true for `Lazy`, because `_` and `(lazy _)` are not equivalent as patterns.